src/lib/ppx_coda/lint_version_syntax.ml

```diff
@@ -0,0 +1,161 @@
+(* lint_version_syntax.ml -- static enforcement of syntactic items relating to proper versioning
+
+   - "deriving bin_io" and "deriving version" never appear in types defined inside functor bodies
+   - otherwise, "bin_io" may appear in a "deriving" attribute only if "version" also appears in that extension
+
+*)
+
+open Core_kernel
+open Ppxlib
+
+let name = "enforce_version_syntax"
+
+let errors_as_warnings_ref = ref false
+
+let is_ident ident item =
+  let version_id id =
+    match id.txt with Lident s -> String.equal s ident | _ -> false
+  in
+  match item with
+  | Pexp_ident id ->
+      version_id id
+  | Pexp_apply ({pexp_desc= Pexp_ident id; _}, _) ->
+      version_id id
+  | _ ->
+      false
+
+let payload_has_item is_item_ident payload =
+  match payload with
+  | PStr structure ->
+      List.exists structure ~f:(fun str ->
+          match str.pstr_desc with
+          | Pstr_eval (expr, _) -> (
+            (* the "ident" can appear as a singleton ident, or in a tuple *)
+            match expr.pexp_desc with
+            | Pexp_ident _ ->
+                is_item_ident expr.pexp_desc
+            | Pexp_apply ({pexp_desc; _}, _) ->
+                is_item_ident pexp_desc
+            | Pexp_tuple items ->
+                List.exists items ~f:(fun item -> is_item_ident item.pexp_desc)
+            | _ ->
+                false )
+          | _ ->
+              false )
+  | _ ->
+      false
+
+let is_version_ident = is_ident "version"
+
+let payload_has_version = payload_has_item is_version_ident
+
+let is_bin_io_ident = is_ident "bin_io"
+
+let payload_has_bin_io = payload_has_item is_bin_io_ident
+
+let attribute_has_deriving_version ((name, payload) : attribute) =
+  String.equal name.txt "deriving" && payload_has_version payload
+
+let attributes_have_deriving_version (attrs : attribute list) =
+  List.exists attrs ~f:attribute_has_deriving_version
+
+let type_has_deriving_version type_decl =
+  attributes_have_deriving_version type_decl.ptype_attributes
+
+let is_deriving name = String.equal name "deriving"
+
+let make_deriving_validator ~pred err_msg type_decl =
+  let get_deriving_error (name, payload) =
+    if is_deriving name.txt then
+      let has_bin_io = payload_has_bin_io payload in
+      let has_version = payload_has_version payload in
+      if pred has_bin_io has_version then Some (name.loc, err_msg) else None
+    else None
+  in
+  List.filter_map type_decl.ptype_attributes ~f:get_deriving_error
+
+let validate_neither_bin_io_nor_version =
+  make_deriving_validator
+    ~pred:(fun has_bin_io has_version -> has_bin_io || has_version)
+    "Deriving bin_io and deriving version disallowed for types in functor body"
+
+let validate_version_if_bin_io =
+  make_deriving_validator
+    ~pred:(fun has_bin_io has_version -> has_bin_io && not has_version)
+    "Must have deriving version if deriving bin_io"
+
+(* traverse AST, collect errors *)
+let check_deriving_usage =
```

Comment:
I still think it might be cleaner to use a separate module expression recursor, and that might even let us go back to using `is_functor` (sorry!). Something like this perhaps? (Apologies for phone typos etc..)

```ocaml
let check_deriving_usage =
  object (self)
    (* bool indicates whether we're in a functor *)
    inherit [bool * (Location.t * string) list] Ast_traverse.fold as super

    method! module_expr me ((is_functor, errors) as acc) =
      match me.pmod_desc with
      | Pmod_functor (_label, _mty, body) ->
          let _, errors = self#module_expr body (true, errors) in
          (is_functor, errors)
      | _ ->
          super#module_expr me acc

    method! structure_item str ((is_functor, errors) as acc) =
      match str.pstr_desc with
      | Pstr_type (_rec_decl, type_decls) ->
          (* for type declaration, check attributes *)
          let f =
            if is_functor then validate_neither_bin_io_nor_version
            else validate_version_if_bin_io
          in
          (is_functor, errors @ List.concat_map ~f type_decls)
      | Pstr_extension ((name, _payload), _attrs)
        when String.equal name.txt "test_module" ->
          (* don't check for errors in test code *)
          (is_functor, errors)
      | _ ->
          super#structure_item str acc
  end
```

Reply:
yes, that would eliminate some of junk cases we see in `structure_item`. Good idea!